aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,644 @@
+    // Reuse table resolver since aggregator values are stored as Table values
+    // internally.
+    resolver: &'a dyn TableResolver,
+    txn_hash: u128,
+    // All existing aggregator tables in this context, wrapped in RefCell for
+    // internal mutability.
+    aggregator_table_data: RefCell<AggregatorTableData>,
+}
+
+impl<'a> NativeAggregatorContext<'a> {
+    /// Creates a new instance of a native aggregator context. This must be
+    /// passed into VM session.
+    pub fn new(txn_hash: u128, resolver: &'a dyn TableResolver) -> Self {
+        Self {
+            resolver,
+            txn_hash,
+            aggregator_table_data: Default::default(),
+        }
+    }
+
+    /// Returns all changes made during this VM session.
+    pub fn into_change_set(self) -> AggregatorChangeSet {
+        let NativeAggregatorContext {
+            aggregator_table_data,
+            ..
+        } = self;
+        let AggregatorTableData { aggregator_tables } = aggregator_table_data.into_inner();
+
+        let mut changes = BTreeMap::new();
+        for (table_handle, aggregator_table) in aggregator_tables {
+            let mut change = BTreeMap::new();
+            let AggregatorTable {
+                removed_aggregators,
+                table,
+                ..
+            } = aggregator_table;
+
+            // Add changes that update or create aggregators.
+            for (key, aggregator) in table {
+                let Aggregator {
+                    state,
+                    value,
+                    limit,
+                    ..
+                } = aggregator;
+
+                let op = match state {
+                    AggregatorState::Data => AggregatorChange::Write(value),
+                    AggregatorState::PositiveDelta => {
+                        let delta_op = DeltaOp::Addition { value, limit };
+                        AggregatorChange::Delta(delta_op)
+                    }
+                };
+                change.insert(key, Some(op));
+            }
+
+            // Add changes that delete aggregators.
+            for key in removed_aggregators {
+                change.insert(key, None);
+            }
+
+            changes.insert(table_handle, change);
+        }
+
+        AggregatorChangeSet { changes }
+    }
+}
+
+// ================================= Natives =================================
+
+/// All aggregator native functions. For more details, refer to code in
+/// `aggregator_table.move`.
+pub fn aggregator_natives(aggregator_addr: AccountAddress) -> NativeFunctionTable {
+    native_functions::make_table(
+        aggregator_addr,
+        &[
+            ("aggregator", "add", native_add),
+            ("aggregator", "read", native_read),
+            ("aggregator", "remove_aggregator", native_remove_aggregator),
+            ("aggregator", "sub", native_sub),
+            ("aggregator_table", "new_aggregator", native_new_aggregator),
+        ],
+    )
+}
+
+fn native_new_aggregator(
```

Comment:
It might be helpful that you paste as a comment the Move signatures of those natives